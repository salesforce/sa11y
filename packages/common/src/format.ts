/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults } from './axe';

export const errMsgHeader = 'Accessibility issues found';

/**
 * Filter to post-process a11y results from axe
 */
export interface Filter {
    (violations: AxeResults, ...args: never[]): AxeResults;
}

// TODO (refactor): constrain rule id to known rule ids e.g using string literal, keyof, in etc
//  e.g. https://stackoverflow.com/a/54061487
// const ruleIDs = getRules().map((ruleObj) => ruleObj.ruleId);
// type RuleID = keyof ruleIDs;
// type RuleID = typeof ruleIDs[number];
// Array of length 2 or greater
type MultiArray<T> = [T, T, ...T[]];

// Selectors within a frame
type BaseSelector = string;

type ShadowDomSelector = MultiArray<BaseSelector>;
type CrossTreeSelector = BaseSelector | ShadowDomSelector;
type RuleID = string;
type CssSelectors = CrossTreeSelector[];
/**
 * Exception list of map of rule to corresponding css targets that needs to be filtered from a11y results.
 */
export type ExceptionList = Record<RuleID, CssSelectors>;
