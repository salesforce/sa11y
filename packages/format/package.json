{
    "name": "@sa11y/format",
    "version": "0.2.1-beta.0",
    "description": "Accessibility results re-formatter",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/format#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/format"
    },
    "keywords": [
        "accessibility",
        "automated testing",
        "axe",
        "results",
        "format",
        "template"
    ],
    "main": "dist/format.js",
    "types": "dist/format.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "dependencies": {
        "axe-core": "3.5.5"
    },
    "devDependencies": {
        "@sa11y/test-utils": "0.2.0-beta.0"
    },
    "publishConfig": {
        "access": "public"
    }
}
