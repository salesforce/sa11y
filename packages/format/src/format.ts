/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { Result } from 'axe-core';

/**
 * Formatter defines the function signature to format accessibility violations found by axe
 */
export interface Formatter {
    (violations: Result[]): string;
}

/**
 * Format accessibility results from axe
 * @param violations - Result list from axe
 * */
// TODO: Add handlebars template for formatting
// TODO: Add support for different output formats console(colored), plain text, HTML, xUnit
export function a11yResultsFormatter(violations: Result[]): string {
    return violations
        .map((violation) => {
            return violation.nodes
                .map((node) => {
                    const selectors = node.target.join(', ');
                    return `${violation.help} (${violation.id}): ${selectors}\n  - More info: ${violation.helpUrl}`;
                })
                .join('\n\n');
        })
        .join('\n\n');
}
