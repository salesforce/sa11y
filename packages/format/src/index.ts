/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yError, Options } from './format';
export { exceptionListFilter } from './filter';
export { A11yResult, A11yResults } from './result';
