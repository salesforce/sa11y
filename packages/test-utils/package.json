{
    "name": "@sa11y/test-utils",
    "version": "0.2.0-beta.0",
    "private": true,
    "description": "Private package providing test utilities for @sa11y packages",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/test-utils#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/test-utils"
    },
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "dependencies": {
        "@sa11y/common": "0.2.0-beta.0"
    }
}
