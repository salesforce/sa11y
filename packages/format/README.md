# `format`

Format accessibility results from axe

## Usage

```
import axe from 'axe-core';
import { a11yResultsFormatter } from '@sa11y/format';

const results = await axe.run();
console.log(a11yResultsFormatter(results.violations));
```
