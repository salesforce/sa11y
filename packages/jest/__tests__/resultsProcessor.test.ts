/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import resultsProcessor from '../src/resultsProcessor';
import { addResult, createEmptyTestResult, makeEmptyAggregatedTestResult } from '@jest/test-result';
import { AggregatedResult, AssertionResult, TestResult } from '@jest/test-result/build/types';
import { A11yError, A11yResult } from '@sa11y/format';
import { getA11yError } from '@sa11y/format/__tests__/format.test';
import { domWithVisualA11yIssues } from '@sa11y/test-utils';
import { expect } from '@jest/globals';

const a11yResults: A11yResult[] = [];
const aggregatedResults = makeEmptyAggregatedTestResult();
const testSuite = createEmptyTestResult();
let numTestFailures = 0;
const numNonA11yFailures = 3;

function addTestFailure(suite: TestResult, err: Error) {
    const failure = {
        // Subset of props used by results processor logic
        failureDetails: [err],
        fullName: `${err.name}-${numTestFailures}`, // Unique test name to test consolidation
        status: 'failed',
        ancestorTitles: ['sa11y'],
    } as AssertionResult;
    suite.testResults.push(failure);
    suite.numFailingTests += 1;
    numTestFailures++;
}

beforeAll(async () => {
    // Prepare test data
    const a11yError = await getA11yError();
    const a11yErrorVisual = await getA11yError(domWithVisualA11yIssues);
    const combinedViolations = [...a11yError.violations, ...a11yErrorVisual.violations];
    a11yResults.push(...a11yError.a11yResults, ...a11yErrorVisual.a11yResults);
    addTestFailure(testSuite, new A11yError(a11yError.violations, a11yError.a11yResults));
    addTestFailure(testSuite, new A11yError(a11yErrorVisual.violations, a11yErrorVisual.a11yResults));
    // Duplicate test result to test consolidation
    addTestFailure(testSuite, new A11yError(a11yError.violations, a11yError.a11yResults));
    addTestFailure(testSuite, new A11yError(a11yErrorVisual.violations, a11yErrorVisual.a11yResults));
    addTestFailure(testSuite, new A11yError(combinedViolations, a11yResults));
    // Add non-a11y test failure
    addTestFailure(testSuite, new Error('foo'));
    testSuite.testFilePath = '/test/data/sa11y-auto-checks.js';
    addResult(aggregatedResults, testSuite);
});

describe('Results Processor', () => {
    it('should have valid test data to start with', () => {
        expect(aggregatedResults.numFailedTestSuites).toBe(1);
        expect(aggregatedResults.numFailedTests).toBe(numTestFailures);
        expect(aggregatedResults).toMatchSnapshot();
    });

    it('should process test results as expected', () => {
        const numA11yFailures = numTestFailures - numNonA11yFailures;
        // Create a copy as results gets mutated by results processor
        const results = JSON.parse(JSON.stringify(aggregatedResults)) as AggregatedResult;
        const processedResults = resultsProcessor(results);
        expect(processedResults).toMatchSnapshot();
        expect(processedResults).not.toEqual(aggregatedResults);
        expect(processedResults.numFailedTestSuites).toEqual(numA11yFailures + 1);
        expect(processedResults.numTotalTests).toEqual(aggregatedResults.numTotalTests + numA11yFailures + 2); // After consolidation + non-a11y failure
    });
});
