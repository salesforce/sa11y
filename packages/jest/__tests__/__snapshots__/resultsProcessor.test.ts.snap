// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results Processor should have valid test data to start with 1`] = `
Object {
  "numFailedTestSuites": 1,
  "numFailedTests": 6,
  "numPassedTestSuites": 0,
  "numPassedTests": 0,
  "numPendingTestSuites": 0,
  "numPendingTests": 0,
  "numRuntimeErrorTestSuites": 0,
  "numTodoTests": 0,
  "numTotalTestSuites": 0,
  "numTotalTests": 6,
  "openHandles": Array [],
  "snapshot": Object {
    "added": 0,
    "didUpdate": false,
    "failure": false,
    "filesAdded": 0,
    "filesRemoved": 0,
    "filesRemovedList": Array [],
    "filesUnmatched": 0,
    "filesUpdated": 0,
    "matched": 0,
    "total": 0,
    "unchecked": 0,
    "uncheckedKeysByFile": Array [],
    "unmatched": 0,
    "updated": 0,
  },
  "startTime": 0,
  "success": true,
  "testResults": Array [
    Object {
      "leaks": false,
      "numFailingTests": 6,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/test/data/sa11y-auto-checks.js",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "sa11y",
          ],
          "failureDetails": Array [
            [A11yError: 5 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/html-has-lang

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one],
          ],
          "fullName": "A11yError-0",
          "status": "failed",
        },
        Object {
          "ancestorTitles": Array [
            "sa11y",
          ],
          "failureDetails": Array [
            [A11yError: 1 Accessibility issues found
 * (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/html-has-lang],
          ],
          "fullName": "A11yError-1",
          "status": "failed",
        },
        Object {
          "ancestorTitles": Array [
            "sa11y",
          ],
          "failureDetails": Array [
            [A11yError: 5 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/html-has-lang

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one],
          ],
          "fullName": "A11yError-2",
          "status": "failed",
        },
        Object {
          "ancestorTitles": Array [
            "sa11y",
          ],
          "failureDetails": Array [
            [A11yError: 1 Accessibility issues found
 * (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/html-has-lang],
          ],
          "fullName": "A11yError-3",
          "status": "failed",
        },
        Object {
          "ancestorTitles": Array [
            "sa11y",
          ],
          "failureDetails": Array [
            [A11yError: 6 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/html-has-lang

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one

* (html-has-lang) <html> element must have a lang attribute: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/html-has-lang],
          ],
          "fullName": "A11yError-4",
          "status": "failed",
        },
        Object {
          "ancestorTitles": Array [
            "sa11y",
          ],
          "failureDetails": Array [
            [Error: foo],
          ],
          "fullName": "Error-5",
          "status": "failed",
        },
      ],
    },
  ],
  "wasInterrupted": false,
}
`;

exports[`Results Processor should process test results as expected 1`] = `
Object {
  "numFailedTestSuites": 5,
  "numFailedTests": 5,
  "numPassedTestSuites": 0,
  "numPassedTests": 0,
  "numPendingTestSuites": 0,
  "numPendingTests": 0,
  "numRuntimeErrorTestSuites": 0,
  "numTodoTests": 0,
  "numTotalTestSuites": 0,
  "numTotalTests": 5,
  "openHandles": Array [],
  "snapshot": Object {
    "added": 0,
    "didUpdate": false,
    "failure": false,
    "filesAdded": 0,
    "filesRemoved": 0,
    "filesRemovedList": Array [],
    "filesUnmatched": 0,
    "filesUpdated": 0,
    "matched": 0,
    "total": 0,
    "unchecked": 0,
    "uncheckedKeysByFile": Array [],
    "unmatched": 0,
    "updated": 0,
  },
  "startTime": 0,
  "success": true,
  "testResults": Array [
    Object {
      "leaks": false,
      "numFailingTests": 1,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/test/data/sa11y-auto-checks.js::SA11Y-WCAG-SC2.4.2-P2",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "sa11y",
            "A11yError-0",
          ],
          "failureDetails": Array [],
          "failureMessages": Array [
            "Accessibility issues found: Documents must have <title> element to aid in navigation
Help: https://dequeuniversity.com/rules/axe/4.3/document-title
CSS Selectors: html
HTML element: <html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>
Tests: \\"A11yError-0\\"
Summary: Fix any of the following:
  Document does not have a non-empty <title> element",
          ],
          "fullName": "Documents must have <title> element to aid in navigation::html",
          "status": "failed",
        },
      ],
    },
    Object {
      "leaks": false,
      "numFailingTests": 1,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/test/data/sa11y-auto-checks.js::SA11Y-WCAG-SC3.1.1-P2",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "sa11y",
            "A11yError-0",
          ],
          "failureDetails": Array [],
          "failureMessages": Array [
            "Accessibility issues found: <html> element must have a lang attribute
Help: https://dequeuniversity.com/rules/axe/4.3/html-has-lang
CSS Selectors: html
HTML element: <html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>
Tests: \\"A11yError-0\\"
Summary: Fix any of the following:
  The <html> element does not have a lang attribute",
          ],
          "fullName": "<html> element must have a lang attribute::html",
          "status": "failed",
        },
      ],
    },
    Object {
      "leaks": false,
      "numFailingTests": 1,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/test/data/sa11y-auto-checks.js::SA11Y-WCAG-SC4.1.2-P1",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "sa11y",
            "A11yError-0",
          ],
          "failureDetails": Array [],
          "failureMessages": Array [
            "Accessibility issues found: Links must have discernible text
Help: https://dequeuniversity.com/rules/axe/4.3/link-name
CSS Selectors: a
HTML element: <a href=\\"#\\"></a>
Tests: \\"A11yError-0\\"
Summary: Fix all of the following:
  Element is in tab order and does not have accessible text

Fix any of the following:
  Element does not have text that is visible to screen readers
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute",
          ],
          "fullName": "Links must have discernible text::a",
          "status": "failed",
        },
      ],
    },
    Object {
      "leaks": false,
      "numFailingTests": 1,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/test/data/sa11y-auto-checks.js::SA11Y-WCAG-SC4.1.1-P3",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "sa11y",
            "A11yError-0",
          ],
          "failureDetails": Array [],
          "failureMessages": Array [
            "Accessibility issues found: Document should have one main landmark
Help: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main
CSS Selectors: html
HTML element: <html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>
Tests: \\"A11yError-0\\"
Summary: Fix all of the following:
  Document does not have a main landmark",
          ],
          "fullName": "Document should have one main landmark::html",
          "status": "failed",
        },
      ],
    },
    Object {
      "leaks": false,
      "numFailingTests": 1,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/test/data/sa11y-auto-checks.js::SA11Y-WCAG-SC1.3.1-P2",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "sa11y",
            "A11yError-0",
          ],
          "failureDetails": Array [],
          "failureMessages": Array [
            "Accessibility issues found: Page should contain a level-one heading
Help: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one
CSS Selectors: html
HTML element: <html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>
Tests: \\"A11yError-0\\"
Summary: Fix all of the following:
  Page must have a level-one heading",
          ],
          "fullName": "Page should contain a level-one heading::html",
          "status": "failed",
        },
      ],
    },
  ],
  "wasInterrupted": false,
}
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 2`] = `4`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 3`] = `
"4 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 1`] = `""`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 2`] = `0`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 3`] = `
"0 Accessibility issues found
 "
`;

exports[`a11y Results Formatter should format using specified options: 0 1`] = `"[{\\"id\\":\\"document-title\\",\\"impact\\":\\"serious\\",\\"tags\\":[\\"cat.text-alternatives\\",\\"wcag2a\\",\\"wcag242\\",\\"ACT\\"],\\"description\\":\\"Ensures each HTML document contains a non-empty <title> element\\",\\"help\\":\\"Documents must have <title> element to aid in navigation\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.3/document-title?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[{\\"id\\":\\"doc-has-title\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Document does not have a non-empty <title> element\\"}],\\"all\\":[],\\"none\\":[],\\"impact\\":\\"serious\\",\\"html\\":\\"<html lang=\\\\\\"en\\\\\\"><head></head><body>\\\\n\\\\n<div id=\\\\\\"dom-with-issues\\\\\\">\\\\n    <a href=\\\\\\"#\\\\\\"></a>\\\\n</div>\\\\n\\\\n\\\\n</body></html>\\",\\"target\\":[\\"html\\"],\\"failureSummary\\":\\"Fix any of the following:\\\\n  Document does not have a non-empty <title> element\\"}]},{\\"id\\":\\"link-name\\",\\"impact\\":\\"serious\\",\\"tags\\":[\\"cat.name-role-value\\",\\"wcag2a\\",\\"wcag412\\",\\"wcag244\\",\\"section508\\",\\"section508.22.a\\",\\"ACT\\"],\\"description\\":\\"Ensures links have discernible text\\",\\"help\\":\\"Links must have discernible text\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.3/link-name?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[{\\"id\\":\\"has-visible-text\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"minor\\",\\"message\\":\\"Element does not have text that is visible to screen readers\\"},{\\"id\\":\\"aria-label\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"aria-label attribute does not exist or is empty\\"},{\\"id\\":\\"aria-labelledby\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\"},{\\"id\\":\\"non-empty-title\\",\\"data\\":{\\"messageKey\\":\\"noAttr\\"},\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element has no title attribute\\"}],\\"all\\":[],\\"none\\":[{\\"id\\":\\"focusable-no-name\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element is in tab order and does not have accessible text\\"}],\\"impact\\":\\"serious\\",\\"html\\":\\"<a href=\\\\\\"#\\\\\\"></a>\\",\\"target\\":[\\"a\\"],\\"failureSummary\\":\\"Fix all of the following:\\\\n  Element is in tab order and does not have accessible text\\\\n\\\\nFix any of the following:\\\\n  Element does not have text that is visible to screen readers\\\\n  aria-label attribute does not exist or is empty\\\\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\\\n  Element has no title attribute\\"}]},{\\"id\\":\\"landmark-one-main\\",\\"impact\\":\\"moderate\\",\\"tags\\":[\\"cat.semantics\\",\\"best-practice\\"],\\"description\\":\\"Ensures the document has a main landmark\\",\\"help\\":\\"Document should have one main landmark\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.3/landmark-one-main?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[],\\"all\\":[{\\"id\\":\\"page-has-main\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"moderate\\",\\"message\\":\\"Document does not have a main landmark\\"}],\\"none\\":[],\\"impact\\":\\"moderate\\",\\"html\\":\\"<html lang=\\\\\\"en\\\\\\"><head></head><body>\\\\n\\\\n<div id=\\\\\\"dom-with-issues\\\\\\">\\\\n    <a href=\\\\\\"#\\\\\\"></a>\\\\n</div>\\\\n\\\\n\\\\n</body></html>\\",\\"target\\":[\\"html\\"],\\"failureSummary\\":\\"Fix all of the following:\\\\n  Document does not have a main landmark\\"}]},{\\"id\\":\\"page-has-heading-one\\",\\"impact\\":\\"moderate\\",\\"tags\\":[\\"cat.semantics\\",\\"best-practice\\"],\\"description\\":\\"Ensure that the page, or at least one of its frames contains a level-one heading\\",\\"help\\":\\"Page should contain a level-one heading\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[],\\"all\\":[{\\"id\\":\\"page-has-heading-one\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"moderate\\",\\"message\\":\\"Page must have a level-one heading\\"}],\\"none\\":[],\\"impact\\":\\"moderate\\",\\"html\\":\\"<html lang=\\\\\\"en\\\\\\"><head></head><body>\\\\n\\\\n<div id=\\\\\\"dom-with-issues\\\\\\">\\\\n    <a href=\\\\\\"#\\\\\\"></a>\\\\n</div>\\\\n\\\\n\\\\n</body></html>\\",\\"target\\":[\\"html\\"],\\"failureSummary\\":\\"Fix all of the following:\\\\n  Page must have a level-one heading\\"}]}]"`;

exports[`a11y Results Formatter should format using specified options: 1 1`] = `
"\\"* (document-title) Documents must have <title> element to aid in navigation: html\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

\\"* (link-name) Links must have discernible text: a\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

\\"* (landmark-one-main) Document should have one main landmark: html\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

\\"* (page-has-heading-one) Page should contain a level-one heading: html\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should format using specified options: 2 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should format using specified options: 3 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should format using specified options: 4 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should not throw error for repeated violations with consolidation 1`] = `
"4 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should not throw error for repeated violations with consolidation 2`] = `
"4 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;

exports[`a11y Results Formatter should throw error when violations are present 1`] = `
"4 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one"
`;
